#python script for my pipeline 

#import fasta file containing the 9000 files 

#use mafft to align sequences in the fasta file above
#and create a new file for these 9000 aligned sequences 

#create a job submission script including a line of raxml or IQ tree 
#to create the maximum likelihood phylogenetic tree 
#check the RAxML_bipartitions file using head and wc -l

#create a python script for the following:
#use biopython to create a file displaying a tree
#read the newick file into a tree object

#root the tree with the longest branch 

#set up figure and axis 

#plot the tree without X and Y axis 

#save plot as a tsv file 
